### PR TITLE
Use new GPU kernel for [unsorted] segment reductions

### PR DESCRIPTION
**Pull request by benbarsdell**
- Optionally replaces the old atomics-based kernels with calls to `SegmentReduceGPU` (the same kernel already used for sparse segment reductions). This behavior is enabled by default, but the old kernels can be re-enabled by setting the environment variable `TF_USE_ATOMIC_SEGMENT_REDUCTIONS=1`. On Windows, the old kernels are always used due to a build issue with the new kernel.
- This improves performance, and guarantees that these ops are deterministic. In future it is hoped that the old kernels can be removed completely.
- Also adds a GPU kernel registration for `SegmentMean`, which didn't previously exist.

cc @nluehr @reedwm @duncanriach

### CODE REVIEW

File: tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc [hunk 1]
```diff
@@ -33,36 +49,32 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
...
-                                          functor::NonAtomicProdOpGpu<T>, \
-                                          functor::AtomicProdOpGpu<T>>;   \
-  template struct SegmentReductionFunctor<T, Index, functor::Highest<T>,  \
-                                          functor::NonAtomicMinOpGpu<T>,  \
-                                          functor::AtomicMinOpGpu<T>>;    \
-  template struct SegmentReductionFunctor<T, Index, functor::Lowest<T>,   \
-                                          functor::NonAtomicMaxOpGpu<T>,  \
-                                          functor::AtomicMaxOpGpu<T>>;
+#define DEFINE_SORTED_GPU_SPECS_INDEX(T, Index)                            \
+  template struct SegmentReductionFunctor<T, Index, functor::Zero<T>,      \
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc:53]: Can you please add `/*EmptySegmentValueF=*/` and `/*InitialValueF=*/` here to make it clear which arg is which?    Same elsewhere.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc [hunk 2]
```diff
@@ -20,6 +20,22 @@ limitations under the License.
 
 namespace tensorflow {
 
+bool UseAtomicSegmentReductions() {
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc]: Done

File: tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/segment_reduction_ops_gpu_0.cu.cc]: Done

**Comment by duncanriach** [comment]
First pass through this. Nice work, @benbarsdell. A couple of thoughts:

1. The original code differentiated between atomic and nonatomic reductions (because there were sometimes both), but not for reasons of determinism. And, of course, atomics can be used in deterministic code. So, there's some lingering old symbology around atomic/nonatomic which has gotten tangled up with (or repurposed for) deterministic/nondeterministic operation. @sanjoy picked out one example of that above. There are other places in the code where perhaps "non-atomic" would ideally be replaced with "deterministic" and where "atomic" would ideally be replaced with "nondeterministic."
2. I'm wondering if we should add tests to confirm determinism, or if we should rely on it being deterministic by design. There will, of course, be backup testing from @reedwm's deterministic auto-checker.

**Comment by benbarsdell** [comment]
Thanks for the suggestions Duncan. I'm facing some build issues right now but will push the additional atomic->deterministic rewordings tomorrow along with a rebase.

I haven't looked closely at how the determinism tests work but will take a look.

**Review by sanjoy — APPROVED** [review]

**Comment by reedwm** [comment]
Rolled forward in 9a1072e69c2bbff89dec51d2bc8042f5a4d0d259

**Comment by reedwm** [comment]
Unfortunately this was rolled back again in 51ee4155408c29a94345dd38595331b189b829fc, since it caused a performance regression in an internal model. Also, the test  `segment_reduction_ops_deterministic_test.py` failed on Windows, since the deterministic algorithms were disabled on Windows.

I think the best approach here is to keep using the old nondeterministic kernels by default, and only use the new ones if either determinism is enabled or an environmental variable is set (say, `TF_USE_DETERMINISTIC_SEGMENT_REDUCTIONS`). Then the determinism API is unblocked and we can internally debug the performance issue, and potentially send you an example to reproduce.

@benbarsdell, do you want to create a new PR with the new kernels disabled by default, by this Wednesday? If not, I can rollforward, disabling the new kernels by default. I want to get this in as soon as possible since it is required for determinism.

**Comment by arneemilhansen** [comment]
Hey all you clever heads! Sorry if this is the wrong place to be commenting but it is directly related to my issue.
Im getting this error:
Deterministic GPU implementation of unsorted segment reduction op not available.
	 [[{{node Adam/Adam/update/UnsortedSegmentSum}}]] [Op:__inference_train_function_15350]

, when trying to fit my model in TF2.10 on Windows with
 `tf.config.experimental.enable_op_determinism()` as well as:
`os.environ["TF_USE_DETERMINISTIC_SEGMENT_REDUCTIONS"]="1"`
I tried to understand whats going on in here:
[https://github.com/benbarsdell/tensorflow/commit/92b294fac35d905f62ac89d66eeaf6e99054f954](url)
Am i understanding it correctly, that this issue has not been solved for Windows so `TF_USE_DETERMINISTIC_SEGMENT_REDUCTIONS` will always be FALSE? 
If so, is there any workaround for Windows users that still gives determinism?

**Pull request merged**